### PR TITLE
Add tests for discussion servlet

### PR DESCRIPTION
**Pull request by n-wach**

### CODE REVIEW

**Comment by n-wach** [comment]
Oops, need to wait for #78 to be merged to shrink changes

**Comment by n-wach** [comment]
Nevermind, I can just change the base branch to `mockito`.

File: src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java [hunk 1]
```diff
@@ -47,8 +47,8 @@
 // TODO: References to Lecture kind string literal need to be updated once Lecture code is merged.
 @WebServlet("/discussion")
 public class DiscussionServlet extends HttpServlet {
-  private static final String PARAM_LECTURE = "lecture";
-  private static final String PARAM_PARENT = "parent";
+  public static final String PARAM_LECTURE = "lecture";
```
Comment (jaysaleh) [inline src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java:50]: Does @visiblefortesting work here?

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    Gson gson = new GsonBuilder().registerTypeAdapterFactory(GenerateTypeAdapter.FACTORY).create();
+    Type listType = new TypeToken<ArrayList<Comment>>() {}.getType();
+    List<Comment> comments = gson.fromJson(json, listType);
+    assertThat(comments.size()).isEqualTo(2);
+  }
+
+  private Entity createTestCommentEntity(int lectureId) {
+    Entity commentEntity = new Entity(Comment.ENTITY_KIND);
+    Key lecture = KeyFactory.createKey(/* kind= */ "Lecture", lectureId);
+    commentEntity.setProperty(Comment.PROP_LECTURE, lecture);
```
Comment (jaysaleh) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:167]: Maybe we should create a util class for Comment Entity consts rather than having them in the AutoValue class. What do you think?

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    assertThat(comments.size()).isEqualTo(3);
+  }
+
+  @Test
+  public void doGet_returnsCommentsForSpecificLecture() throws ServletException, IOException {
+    request.setParameter(DiscussionServlet.PARAM_LECTURE, "2");
+    datastore.put(createTestCommentEntity(1));
+    datastore.put(createTestCommentEntity(1));
+    datastore.put(createTestCommentEntity(1));
+    datastore.put(createTestCommentEntity(2)); // this one
```
Comment (jaysaleh) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:149]: These comments are confusing

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+  }
+
+  @Test
+  public void doGet_returnsNothingForUnknownLecture() throws ServletException, IOException {
+    request.setParameter(DiscussionServlet.PARAM_LECTURE, "1");
+
+    servlet.doGet(request, response);
+
+    assertThat(response.getContentType()).isEqualTo("application/json;");
+    assertThat(response.getContentAsString()).startsWith("[]");
```
Comment (jaysaleh) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:123]: I came across this as well, maybe we should add a comment as to why we do `startsWith("[]")`. I can see someone who hasn't dealt with the code being confused here

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    datastore.put(createTestCommentEntity(1));
+    datastore.put(createTestCommentEntity(1));
+    datastore.put(createTestCommentEntity(1));
+
+    servlet.doGet(request, response);
+
+    assertThat(response.getContentType()).isEqualTo("application/json;");
+    String json = response.getContentAsString();
+    Gson gson = new GsonBuilder().registerTypeAdapterFactory(GenerateTypeAdapter.FACTORY).create();
+    Type listType = new TypeToken<ArrayList<Comment>>() {}.getType();
```
Comment (jaysaleh) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:138]: Up to you, this code is used more than once, maybe make a helper function to clean up the test

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    servlet.doPost(request, response);
+
+    assertThat(response.getStatus()).isEqualTo(HttpServletResponse.SC_FORBIDDEN);
+    assertThat(datastore.prepare(new Query(Comment.ENTITY_KIND)).countEntities(withLimit(1)))
+        .isEqualTo(0);
+  }
+
+  @Test
+  public void doPost_storesCommentWithProperties_noParent() throws ServletException, IOException {
+    int LECTURE_ID = 1;
```
Comment (amuamushu) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:74]: Can you structure your tests to be arrange, act, assert?

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+    assertThat(comment.lecture().getId()).isEqualTo(LECTURE_ID);
+    assertThat(comment.author().getEmail()).isEqualTo(EMAIL);
+    assertThat(comment.content()).isEqualTo(CONTENT);
+    assertThat(comment.parent().isPresent()).isFalse();
+  }
+
+  @Test
+  public void doPost_storesCommentParent() throws ServletException, IOException {
+    int PARENT_ID = 32;
+    testServices.setEnvIsLoggedIn(true);
```
Comment (amuamushu) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:99]: arrange, act, assert here too please

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+
+    assertThat(response.getStatus()).isEqualTo(HttpServletResponse.SC_ACCEPTED);
+    Comment comment = Comment.fromEntity(query.asSingleEntity());
+    assertThat(comment.parent().isPresent()).isTrue();
+    assertThat(comment.parent().get().getId()).isEqualTo(PARENT_ID);
+  }
+
+  @Test
+  public void doGet_returnsNothingForUnknownLecture() throws ServletException, IOException {
+    request.setParameter(DiscussionServlet.PARAM_LECTURE, "1");
```
Comment (amuamushu) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:118]: Should 1 be a constant variable? I feel having the value itself is a bit distracting since I'm not sure what the significance of 1 is. Also, I noticed you used 1 in your other tests so I think it would be nice to define it globally.

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+    String json = response.getContentAsString();
+    Gson gson = new GsonBuilder().registerTypeAdapterFactory(GenerateTypeAdapter.FACTORY).create();
+    Type listType = new TypeToken<ArrayList<Comment>>() {}.getType();
+    List<Comment> comments = gson.fromJson(json, listType);
+    assertThat(comments.size()).isEqualTo(3);
+  }
+
+  @Test
+  public void doGet_returnsCommentsForSpecificLecture() throws ServletException, IOException {
+    request.setParameter(DiscussionServlet.PARAM_LECTURE, "2");
```
Comment (amuamushu) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:145]: Can you make 2 a constant? I'm not sure what 2 signifies.

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+    datastore.put(createTestCommentEntity(1));
+    datastore.put(createTestCommentEntity(1));
```
Comment (amuamushu) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:147]: I don't think it's clear that `createTestCommentEntity` takes the lecture id as the parameter just from reading the test. Could you either update the method header to use constants instead of numbers?

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+    Comment comment = Comment.fromEntity(query.asSingleEntity());
+    assertThat(comment.lecture().getId()).isEqualTo(LECTURE_ID);
+    assertThat(comment.author().getEmail()).isEqualTo(EMAIL);
+    assertThat(comment.content()).isEqualTo(CONTENT);
+    assertThat(comment.parent().isPresent()).isFalse();
+  }
+
+  @Test
+  public void doPost_storesCommentParent() throws ServletException, IOException {
+    int PARENT_ID = 32;
```
Comment (amuamushu) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:98]: Should this be camel case? It's not a final variable.

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+
+  @After
+  public void cleanUp() {
+    testServices.tearDown();
+  }
+
+  @Test
+  public void doPost_loggedOut_postForbidden() throws ServletException, IOException {
+    testServices.setEnvIsLoggedIn(false);
+    String LECTURE_ID = "1";
```
Comment (amuamushu) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:62]: Since LECTURE_ID is defined in multiple tests, can you create a constant for them? The 1 is a bit distracting, assuming it doesn't mean anything.

File: src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java (hunk 1 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java:50]: IDK.  Gonna wait until we figure out if we're using a util/consts class for all of these

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 3 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:149]: Should be clearer now

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 4 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:123]: I just parsed the json into a list and made sure it was empty

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 5 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:138]: Created helper function

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 6 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:74]: Done

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 7 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:99]: Done

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 8 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:118]: I added some constants, should be clearer now

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 9 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:145]: I added some constants, should be clearer now

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 10 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:147]: I added some constants, should be clearer now

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 11 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:98]: Added it as a constant

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 12 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:62]: I added some constants, should be clearer now

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 2 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:186]: We'll talk about organizing these soon

File: src/test/java/com/googleinterns/zoomtube/servlets/AuthenticationServletTest.java [hunk 13]
```diff
@@ -81,11 +82,7 @@ public void doGet_loggedIn_expectFalse() throws Exception {
 
   @Test
   public void doGet_returnsUserEmail() throws Exception {
-    final String EMAIL = "test@example.com";
-    authService.setEnvIsLoggedIn(true);
-    authService.setEnvAuthDomain("example.com");
-    authService.setEnvEmail(EMAIL);
```
Comment (billie-cat) [inline src/test/java/com/googleinterns/zoomtube/servlets/AuthenticationServletTest.java:87]: This is an important part of the test and therefore should be done here. If someone was reading these tests they would need to know that they need to look at setup to find where the email was set.

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+  private static final int LECTURE_B = 2;
+  private static final int LECTURE_C = 3;
+  private static final String LECTURE_A_STR = "1";
+  private static final String LECTURE_B_STR = "2";
+  private static final String LECTURE_C_STR = "3";
+  private static final int PARENT_ID = 32;
+  private static final String PARENT_ID_STR = "32";
+  private static final String CONTENT = "Test content";
+  private static final String EMAIL = "test@example.com";
+  private static final String AUTH_DOMAIN = "example.com";
```
Comment (billie-cat) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:58]: I thought we had talked about keeping these in the tests themselves to reduce cognitive load? Why did you move them all up here?

File: src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java (hunk 1 above)
Comment (billie-cat) [reply src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java:50]: So these need to be public because of tests? In that case I would make these private and just duplicate the variables in the tests. The issue with making these public is that you end up allowing other classes to use them. 

+1 to what Jay said. I would try @VisibleForTesting because duplicating the variable in tests.

File: src/test/java/com/googleinterns/zoomtube/servlets/AuthenticationServletTest.java (hunk 13 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/AuthenticationServletTest.java:87]: Duplicated to relevant test

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 14 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:52]: Talked in chat, made changes.

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+import org.mockito.junit.MockitoRule;
+
+public class DiscussionServletTest {
+  private static final int LECTURE_ID = 1;
+  private static final String LECTURE_ID_STR = "1";
+  private static final String TEST_COMMENT_CONTENT = "Test content";
+  private static final String DEFAULT_EMAIL = "test@example.com";
+  private static final String AUTH_DOMAIN = "example.com";
+
+  @Rule public final MockitoRule mockito = MockitoJUnit.rule();
```
Comment (f-camp) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:54]:
```suggestion
  @Rule private final MockitoRule mockito = MockitoJUnit.rule();
```

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+    assertThat(comment.lecture().getId()).isEqualTo(LECTURE_ID);
+    assertThat(comment.author().getEmail()).isEqualTo("author@example.com");
+    assertThat(comment.content()).isEqualTo("Something unique");
+    assertThat(comment.parent().isPresent()).isFalse();
+  }
+
+  @Test
+  public void doPost_storesCommentParent() throws ServletException, IOException {
+    testServices.setEnvIsLoggedIn(true);
+    final int PARENT_ID = 32;
```
Comment (f-camp) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:111]:
```suggestion
   final int parentId = 32;
```

is fine here.

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+    verify(response).setContentType("application/json;");
+    List<Comment> comments = getCommentsFromJson(content.toString());
+    assertThat(comments.size()).isEqualTo(3);
+  }
+
+  @Test
+  public void doGet_returnsCommentsForSpecificLecture() throws ServletException, IOException {
+    final int LECTURE_A = 1;
+    final int LECTURE_B = 2;
+    final int LECTURE_C = 3;
```
Comment (f-camp) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:162]: ditto

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 15 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:54]: `org.junit.internal.runners.rules.ValidationError: The @Rule 'mockito' must be public.`

Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:54]: No idea why, or if this can be fixed, but I can't run with it private

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 16 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java]: Fixed!

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 17 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java]: Fixed!

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+import org.junit.Rule;
+import org.junit.Test;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+
+public class DiscussionServletTest {
+  private static final int LECTURE_ID = 1;
+  private static final String LECTURE_ID_STR = "1";
+  private static final String TEST_COMMENT_CONTENT = "Test content";
```
Comment (billie-cat) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:50]: nit: could probably rename this to COMMENT_CONTENT

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+  }
+
+  @Test
+  public void doPost_loggedOut_postForbidden() throws ServletException, IOException {
+    testServices.setEnvIsLoggedIn(false);
+    when(request.getParameter(DiscussionServlet.PARAM_LECTURE)).thenReturn(LECTURE_ID_STR);
+
+    servlet.doPost(request, response);
+
+    verify(response).sendError(eq(HttpServletResponse.SC_FORBIDDEN), anyString());
```
Comment (billie-cat) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:85]: It is better to actually add the string you expect to return. The more specific the test the better.

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+    when(request.getParameter(DiscussionServlet.PARAM_LECTURE)).thenReturn(LECTURE_ID_STR);
+    StringWriter content = new StringWriter();
+    PrintWriter writer = new PrintWriter(content);
+    when(response.getWriter()).thenReturn(writer);
+
+    servlet.doGet(request, response);
+
+    verify(response).setContentType("application/json;");
+    List<Comment> comments = getCommentsFromJson(content.toString());
+    assertThat(comments.isEmpty()).isTrue();
```
Comment (billie-cat) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:139]: you can rewrite it as: assertThat(comments).isEmpty();

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+    datastore.put(createTestCommentEntity(LECTURE_ID));
+    StringWriter content = new StringWriter();
+    PrintWriter writer = new PrintWriter(content);
+    when(response.getWriter()).thenReturn(writer);
+
+    servlet.doGet(request, response);
+
+    verify(response).setContentType("application/json;");
+    List<Comment> comments = getCommentsFromJson(content.toString());
+    assertThat(comments.size()).isEqualTo(3);
```
Comment (billie-cat) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:156]: you can use: assertThat(comments).hasSize(3);

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 22]
```diff
@@ -0,0 +1,205 @@
...
+  }
+
+  private Entity createTestCommentEntity(int lectureId) {
+    Entity commentEntity = new Entity(Comment.ENTITY_KIND);
+    Key lecture = KeyFactory.createKey(/* kind= */ "Lecture", lectureId);
+    commentEntity.setProperty(Comment.PROP_LECTURE, lecture);
+    commentEntity.setProperty(Comment.PROP_PARENT, null);
+    commentEntity.setProperty(Comment.PROP_TIMESTAMP, new Date(0));
+    commentEntity.setProperty(Comment.PROP_AUTHOR, new User(DEFAULT_EMAIL, AUTH_DOMAIN));
+    commentEntity.setProperty(Comment.PROP_CONTENT, TEST_COMMENT_CONTENT);
```
Comment (billie-cat) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:193]: Is this data related to the other test which also uses TEST_COMMENT_CONTENT? I would suggest not using the same value as it adds confusion as to if they are related. If they are related then it should be more obvious. Eg, you can pass in this field.

Comment (billie-cat) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:193]: Same thing for DEFAULT_EMAIL and AUTH_DOMAIN

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 23]
```diff
@@ -0,0 +1,205 @@
...
+    commentEntity.setProperty(Comment.PROP_CREATED, new Date(Clock.systemUTC().millis()));
+
+    return commentEntity;
+  }
+
+  private List<Comment> getCommentsFromJson(String json) {
+    Gson gson = new GsonBuilder().registerTypeAdapterFactory(GenerateTypeAdapter.FACTORY).create();
+    Type listType = new TypeToken<ArrayList<Comment>>() {}.getType();
+    List<Comment> comments = gson.fromJson(json, listType);
```
Comment (billie-cat) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:202]: can you do: return gson.fromJson(json, listType);

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 24]
```diff
@@ -0,0 +1,205 @@
...
+
+    servlet.doPost(request, response);
+
+    verify(response).sendError(eq(HttpServletResponse.SC_FORBIDDEN), anyString());
+    assertThat(datastore.prepare(new Query(Comment.ENTITY_KIND)).countEntities(withLimit(1)))
+        .isEqualTo(0);
+  }
+
+  @Test
+  public void doPost_storesCommentWithProperties_noParent() throws ServletException, IOException {
```
Comment (billie-cat) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:91]: This test name confuses me. What are you testing? Seems you are testing 2 things, storing comments and not having any parents. Should it be split into 2 tests?

File: src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java [hunk 25]
```diff
@@ -47,8 +48,8 @@
 // TODO: References to Lecture kind string literal need to be updated once Lecture code is merged.
 @WebServlet("/discussion")
 public class DiscussionServlet extends HttpServlet {
-  private static final String PARAM_LECTURE = "lecture";
-  private static final String PARAM_PARENT = "parent";
+  @VisibleForTesting protected static final String PARAM_LECTURE = "lecture";
```
[the remark this answers is not in the document]
Comment (f-camp) [reply src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java:51]: why don't we just make it public instead? Since it is used outside of this class. We should be conservative with @VisibleForTesting and find other options before opting for it.

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 24 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:91]: I check every property except the parent property, and test the parent property in another test.  Maybe I should combine them?

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 23 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:202]: changed

File: src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java:51]: Changed

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 26]
```diff
@@ -0,0 +1,205 @@
...
+  private static final String TEST_COMMENT_CONTENT = "Test content";
+  private static final String DEFAULT_EMAIL = "test@example.com";
+  private static final String AUTH_DOMAIN = "example.com";
+
+  @Rule public final MockitoRule mockito = MockitoJUnit.rule();
+
+  @Mock private HttpServletRequest request;
+  @Mock private HttpServletResponse response;
+  private DiscussionServlet servlet;
+  private LocalServiceTestHelper testServices = new LocalServiceTestHelper(
```
[the remark this answers is not in the document]
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:59]: Changed

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 27]
```diff
@@ -0,0 +1,205 @@
...
+  @Test
+  public void doPost_loggedOut_postForbidden() throws ServletException, IOException {
+    testServices.setEnvIsLoggedIn(false);
+    when(request.getParameter(DiscussionServlet.PARAM_LECTURE)).thenReturn(LECTURE_ID_STR);
+
+    servlet.doPost(request, response);
+
+    verify(response).sendError(eq(HttpServletResponse.SC_FORBIDDEN), anyString());
+    assertThat(datastore.prepare(new Query(Comment.ENTITY_KIND)).countEntities(withLimit(1)))
+        .isEqualTo(0);
```
[the remark this answers is not in the document]
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:87]: We can only get the count, an iteratable, or a single entity (which can be null if there is none).  I think this is most clear.  Also, isEmpty doesn't seem to be defined in my version of google truth?

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 21 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:156]: Not supported by this version of truth

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 20 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:139]: Not supported by this version of truth

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 27 above)
[the remark this answers is not in the document]
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:87]: Not supported by this version of truth

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 24 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:91]: I moved things around so it's more clear what I'm testing

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 22 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:193]: I removed these constants as there were only used in one place and their values were never tested anywhere,

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 18 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:50]: Removed it

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 19 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java]: Done

File: src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (billie-cat) [reply src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java]: @f-camp isn't making it public because of tests less ideal than @VisibleForTesting?

File: src/test/java/com/googleinterns/zoomtube/servlets/AuthenticationServletTest.java [hunk 28]
```diff
@@ -29,26 +29,28 @@
   private AuthenticationServlet servlet;
   @Mock private HttpServletRequest request;
   @Mock private HttpServletResponse response;
```
Comment (f-camp) [inline src/test/java/com/googleinterns/zoomtube/servlets/AuthenticationServletTest.java:31]: nit: It is good practice to have mocks at the beginning of tests.

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 29]
```diff
@@ -0,0 +1,204 @@
...
+public class DiscussionServletTest {
+  private static final int LECTURE_ID = 1;
+  private static final String LECTURE_ID_STR = "1";
+  private static final LocalServiceTestHelper testServices = new LocalServiceTestHelper(
+      new LocalUserServiceTestConfig(), new LocalDatastoreServiceTestConfig());
+
+  @Rule public final MockitoRule mockito = MockitoJUnit.rule();
+
+  @Mock private HttpServletRequest request;
+  @Mock private HttpServletResponse response;
```
Comment (f-camp) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:56]: nit: it is good practice to have @Rule and @Mocks at the beginning of tests.

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java [hunk 30]
```diff
@@ -0,0 +1,204 @@
...
+  private DatastoreService datastore;
+
+  @Before
+  public void setUp() throws ServletException {
+    testServices.setUp();
+    testServices.setEnvEmail("test@example.com");
+    testServices.setEnvAuthDomain("example.com");
+    servlet = new DiscussionServlet();
+    servlet.init();
+    datastore = DatastoreServiceFactory.getDatastoreService();
```
Comment (f-camp) [inline src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:67]: Is it the same datastore here that is used form within the servlet init as well?

Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:67]: Because I created a new LocalServiceTestHelper, any call to .getDatastoreService will return the test service.  And because I .setUp before calling servlet.init() (which does .getDatastoreService()), both the servlet and the tests will be using the same test datastore.

File: src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java (hunk 29 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/DiscussionServletTest.java:56]: Moved

File: src/test/java/com/googleinterns/zoomtube/servlets/AuthenticationServletTest.java (hunk 28 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/AuthenticationServletTest.java:31]: Moved
